smart_open/ssh.py

```diff
@@ -0,0 +1,81 @@
+"""Implements I/O streams over SSH.
+
+Examples
+--------
+
+>>> with open('/proc/version_signature', host='1.2.3.4') as conn:
+...     print(conn.read())
+b'Ubuntu 4.4.0-1061.70-aws 4.4.131'
+
+Similarly, from a command line::
+
+    $ python -c "from smart_open import ssh;print(ssh.open('/proc/version_signature', host='1.2.3.4').read())"
+    b'Ubuntu 4.4.0-1061.70-aws 4.4.131'
+
+"""
+
+import getpass
+import logging
+logger = logging.getLogger(__name__)
+
+try:
+    import paramiko
+except ImportError:
+    logger.warning('paramiko missing, SSH will be disabled.  `pip install paramiko` to suppress')
+
+#
+# Global storage for SSH connections.
+#
+_SSH = {}
```

Comment:
How does such global cache play with multiprocessing / forking?

Reply:
For multiprocessing: no problem, there will be one cache for subprocess.

For forking: I think it will be a problem. We don't want different threads to access the same connection simultaneously. Perhaps we can include the current thread_id in the cache key?

Reply:
I don't think we have any solid contract in place in `smart_open` (yet). My naive expectation (as a user) would be "a connection open in one process/thread is always valid for that process/tread, but not necessarily across fork boundaries".

How do the other connection types (S3, HTTP, HDFS…) manage in the face of multiple processes or threads?

Reply:
OTOH they "should be fine". We haven't had any concurrency-related bug reports.

On the other hand, we don't have any explicit testing for concurrency problems, either.